airbyte-server/src/main/java/io/airbyte/server/converters/DatabaseArchiver.java

```diff
@@ -0,0 +1,138 @@
+ */
+
+package io.airbyte.server.converters;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.commons.lang.CloseableConsumer;
+import io.airbyte.commons.lang.Exceptions;
+import io.airbyte.commons.yaml.Yamls;
+import io.airbyte.db.Database;
+import io.airbyte.db.ExceptionWrappingDatabase;
+import io.airbyte.validation.json.JsonSchemaValidator;
+import io.airbyte.validation.json.JsonValidationException;
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.jooq.JSONFormat;
+import org.jooq.JSONFormat.RecordFormat;
+import org.jooq.Record;
+import org.jooq.Table;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DatabaseArchiver {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DatabaseArchiver.class);
+  private static final String DB_FOLDER_NAME = "airbyte_db";
+  private static final String DEFAULT_SCHEMA = "public";
+  private static final JSONFormat DB_JSON_FORMAT = new JSONFormat().recordFormat(RecordFormat.OBJECT);
+
+  private final ExceptionWrappingDatabase database;
+  private final Path storageRoot;
+  private final JsonSchemaValidator jsonSchemaValidator;
+
+  public DatabaseArchiver(final Database database, final Path storageRoot, final JsonSchemaValidator jsonSchemaValidator) {
+    this.database = new ExceptionWrappingDatabase(database);
+    this.storageRoot = storageRoot;
+    this.jsonSchemaValidator = jsonSchemaValidator;
+  }
+
+  public DatabaseArchiver(final Database database, final Path storageRoot) {
+    this(database, storageRoot, new JsonSchemaValidator());
+  }
+
+  /**
+   * Serializes each internal Airbyte Database table into a single archive file stored in YAML.
+   */
+  public void writeDatabaseToArchive() throws IOException {
+    List<Table<?>> tables = listTables();
+    if (tables != null) {
+      tables.forEach(table -> Exceptions.toRuntime(() -> writeTableToArchive(table)));
+    }
+    LOGGER.debug("Successful export of airbyte database");
+  }
+
+  private List<Table<?>> listTables() throws IOException {
+    // list tables from public schema only
+    return database.query(context -> context.meta().getSchemas(DEFAULT_SCHEMA).stream()
+        .flatMap(schema -> context.meta(schema).getTables().stream())
+        .collect(Collectors.toList()));
+  }
+
+  private void writeTableToArchive(Table<?> table) throws Exception {
+    final Path tablePath = buildTablePath(table.getName());
+    Files.createDirectories(tablePath.getParent());
+    final BufferedWriter recordOutputWriter = new BufferedWriter(new FileWriter(tablePath.toFile()));
+    final CloseableConsumer<JsonNode> recordConsumer = Yamls.listWriter(recordOutputWriter);
+    final Stream<Record> records = database.query(ctx -> ctx.select(table.fields())
```

Comment:
i would suggest using java try-with-resource here so that we have a better guarantee that the stream actually gets closed. something like this :
```
    try(final Stream<Record> records = database.query(ctx -> ctx.select(table.fields()).from(table.getName()).fetchStream())) {
    records.forEach(r -> Exceptions.toRuntime(() -> {
      final JsonNode row = Jsons.deserialize(r.formatJSON(DB_JSON_FORMAT));
      // TODO validate table schemas before writing?
      recordConsumer.accept(row);
    }));
    }
```